Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -94,4 +102,24 @@ void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& ori
     }
 }
 
+void WebPermissionController::permissionMayHaveChanged(WebCore::PermissionName permissionName, const WebCore::SecurityOriginData& topOrigin)
+{
+    for (auto& observer : m_observers) {
+        if (observer.origin().topOrigin == topOrigin && observer.descriptor().name == permissionName) {
+            auto& context = observer.context();
+            auto source = WebCore::Permissions::sourceFromContext(context);
+            // FIXME: Add support for service workers and shared workers.
+            ASSERT(source && (*source == WebCore::PermissionQuerySource::Window || *source == WebCore::PermissionQuerySource::DedicatedWorker));
+            auto* page = downcast<WebCore::Document>(context).page();
+
+            if (page) {
+                query(WebCore::ClientOrigin { topOrigin, topOrigin }, WebCore::PermissionDescriptor { permissionName }, page, *source, [&observer](auto newState) {
```

Comment:
I don't believe using topOrigin twice is correct here. You need to use the correct frame origin for the second member, which would likely be `downcast<WebCore::Document>(context)->securityOrigin()`